python/message.c

```diff
@@ -1031,34 +1033,50 @@ static PyObject* PyUpb_CMessage_Clear(PyUpb_CMessage* self, PyObject* args) {
   Py_RETURN_NONE;
 }
 
-static void PyUpb_CMessage_AbandonField(PyUpb_CMessage* self,
-                                        const upb_fielddef* f) {
-  if (self->unset_subobj_map && upb_fielddef_issubmsg(f) &&
-      !upb_fielddef_isseq(f)) {
-    PyObject* sub = PyUpb_WeakMap_Get(self->unset_subobj_map, f);
-    if (sub) {
-      PyUpb_CMessage_AssureReified((PyUpb_CMessage*)sub);
+void PyUpb_CMessage_DoClearField(PyObject* _self, const upb_fielddef* f) {
+  PyUpb_CMessage* self = (void*)_self;
+  PyUpb_CMessage_AssureReified((PyUpb_CMessage*)self);
+
+  if (upb_fielddef_ismap(f)) {
+    // We have to invalidate any existing iterator over this map.
+    PyObject* obj = NULL;
+    if (self->unset_subobj_map) {
+      obj = PyUpb_WeakMap_Get(self->unset_subobj_map, f);
+    }
+    if (!obj) {
+      upb_msg* msg = PyUpb_CMessage_GetMsg(self);
```

Comment:
please add a comment here saying even if we were unable to get it ourselves, we may have a cached version that we still need to invalidate

Reply:
I realized that this code could be clarified and tested better. PTAL.